saexec/execution.go

```diff
@@ -0,0 +1,193 @@
+
+	"github.com/ava-labs/strevm/blocks"
+	"github.com/ava-labs/strevm/dummy"
+	"github.com/ava-labs/strevm/hook"
+)
+
+var errExecutorClosed = errors.New("saexec.Executor closed")
+
+// Enqueue pushes a new block to the FIFO queue. If [Executor.Close] is called
+// before [blocks.Block.Executed] returns true then there is no guarantee that
+// the block will be executed.
+func (e *Executor) Enqueue(ctx context.Context, block *blocks.Block) error {
+	for {
+		select {
+		case e.queue <- block:
+			return nil
+		case <-e.quit:
+			return errExecutorClosed
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			// If this happens then increase the channel's buffer size.
+			e.log.Warn("Execution queue buffer too small")
+			runtime.Gosched()
+		}
+	}
+}
+
+func (e *Executor) processQueue() {
+	defer close(e.done)
+
+	for {
+		select {
+		case <-e.quit:
+			return
+
+		case block := <-e.queue:
+			logger := e.log.With(
+				zap.Uint64("block_height", block.Height()),
+				zap.Uint64("block_time", block.BuildTime()),
+				zap.Any("block_hash", block.Hash()),
+				zap.Int("tx_count", len(block.Transactions())),
+			)
+
+			if err := e.execute(block, logger); err != nil {
+				logger.Error("Block execution failed", zap.Error(err))
+				return
+			}
+		}
+	}
+}
+
+type executionScratchSpace struct {
+	snaps   *snapshot.Tree
+	statedb *state.StateDB
+}
+
+func (e *Executor) execute(b *blocks.Block, logger logging.Logger) error {
+	logger.Debug("Executing block")
+
+	// If the VM were to encounter an error after enqueuing the block, we would
+	// receive the same block twice for execution should consensus retry
+	// acceptance.
+	if last, curr := e.lastExecuted.Load().Height(), b.Height(); curr != last+1 {
+		return fmt.Errorf("executing blocks out of order: %d then %d", last, curr)
+	}
+
+	scratch := &e.executeScratchSpace
+	rules := e.chainConfig.Rules(b.Number(), true /*isMerge*/, b.BuildTime())
+
+	if err := hook.BeforeBlock(e.hooks, rules, scratch.statedb, b, e.gasClock); err != nil {
+		return fmt.Errorf("before-block hook: %v", err)
+	}
+	perTxClock := e.gasClock.Time.Clone()
+
+	header := types.CopyHeader(b.Header())
+	header.BaseFee = e.gasClock.BaseFee().ToBig()
+
+	gasPool := core.GasPool(math.MaxUint64) // required by geth but irrelevant so max it out
+	var blockGasConsumed gas.Gas
+
+	receipts := make(types.Receipts, len(b.Transactions()))
+	for ti, tx := range b.Transactions() {
+		scratch.statedb.SetTxContext(tx.Hash(), ti)
+
+		receipt, err := core.ApplyTransaction(
+			e.chainConfig,
+			dummy.ChainContext(),
+			&header.Coinbase,
+			&gasPool,
+			scratch.statedb,
+			header,
+			tx,
+			(*uint64)(&blockGasConsumed),
+			vm.Config{},
+		)
+		if err != nil {
+			// This almost certainly means that the worst-case block inclusion
+			// has a bug.
+			logger.Error(
+				"Transaction execution errored (not reverted)",
+				zap.Int("tx_index", ti),
+				zap.Stringer("tx_hash", tx.Hash()),
+				zap.Error(err),
+			)
+			continue
```

Comment:
Should this `return err`? By continuing here, we are getting into a situation where we could be including transactions on chain that weren't able to process correctly.

Admittedly, if we are in this position we are in a really bad spot... But wouldn't we want to limit the amount of state corruption as much as possible?